src/coreclr/pal/src/libunwind/include/libunwind-arm.h

```diff
@@ -269,32 +269,42 @@ typedef struct unw_tdep_context
-   use cases.  Stores pc+8, which is only approximately correct, really.  */
+/* FIXME: this is a stub version which only saves GP registers.  Not ideal, but
+   may be sufficient for all libunwind use cases.
+   In thumb mode, we return directly back to thumb mode on return (with bx), to
+   avoid altering any registers after unw_resume. */
 #ifndef __thumb__
-#define unw_tdep_getcontext(uc) (({                                     \
+#define unw_tdep_getcontext(uc) ({                                      \
   unw_tdep_context_t *unw_ctx = (uc);                                   \
-  register unsigned long *unw_base __asm__ ("r0") = unw_ctx->regs;      \
+  register unsigned long *r0 __asm__ ("r0");                            \
+  register unsigned long *unw_base __asm__ ("r1") = unw_ctx->regs;      \
   __asm__ __volatile__ (                                                \
-    "stmia %[base], {r0-r15}"                                           \
-    "adds %[base], #64"                                                 \
-    "vstmia %[base], {d0-d15}"                                          \
-    : : [base] "r" (unw_base) : "memory");                              \
-  }), 0)
+    "mov r0, #0\n"                                                      \
+    "stmia %[base], {r0-r15}\n"                                         \
+    "nop\n" /* align return address to value stored by stmia */         \
+    : [r0] "=r" (r0) : [base] "r" (unw_base) : "memory");               \
+  (int)r0; })
 #else /* __thumb__ */
-#define unw_tdep_getcontext(uc) (({                                     \
+#define unw_tdep_getcontext(uc) ({                                      \
   unw_tdep_context_t *unw_ctx = (uc);                                   \
-  register unsigned long *unw_base __asm__ ("r0") = unw_ctx->regs;      \
+  register unsigned long *r0 __asm__ ("r0");                            \
+  register unsigned long *unw_base __asm__ ("r1") = unw_ctx->regs;      \
   __asm__ __volatile__ (                                                \
-    ".align 2\nbx pc\nnop\n.code 32\n"                                  \
-    "stmia %[base], {r0-r15}\n"                                         \
-    "adds %[base], #64\n"                                               \
-    "vstmia %[base], {d0-d15}\n"                                        \
-    "orr %[base], pc, #1\nbx %[base]\n"                                 \
-    ".code 16\n"							\
-    : [base] "+r" (unw_base) : : "memory", "cc");                       \
-  }), 0)
+    ".align 2\n"                                                        \
+    "bx pc\n"                                                           \
+    "nop\n"                                                             \
+    ".code 32\n"                                                        \
+    "mov r0, #0\n"                                                      \
+    "stmia %[base], {r0-r14}\n"                                         \
```

Comment:
This is now missing my changes that added storing the floating point registers. In both the thumb and non-thumb code.   For the `#else` branch, just adding my two lines after the `stmia` should be enough. For the `#if` branch, we would need to make sure that the PC stored by the `stmia` matches the new semantics of the code from the libunwind repo, that means that it points after the `nop`. It seems that the easiest is to store the floating point registers first and then the floating point ones. A simple approach would be to add the following before the `stmia`:
```
    "adds %[base], #64\n" \
    "vstmia %[base], {d0-d15}\n" \
    "subs %[base], #64\n" \
```

Reply:
Applied and rebased. All arm changes are in a single commit: 0ca7809.

Reply:
@vtjnash You mean doing that and also adding `!` to the `stmia %[base], {r0-r15}` => `stmia %[base]!, {r0-r15}`, right? Yes, that's nice.
Btw, I've just found that the offset of the stored PC is actually implementation-defined constant. Here the code from the upstream libunwind assumes it is 8.

Reply:
@am11, based on @vtjnash's suggestion, can you please update the patch to
```diff
diff --git a/src/coreclr/pal/src/libunwind/include/libunwind-arm.h b/src/coreclr/pal/src/libunwind/include/libunwind-arm.h
index 9108dd9d9a3..c5c2b3072ac 100644
--- a/src/coreclr/pal/src/libunwind/include/libunwind-arm.h
+++ b/src/coreclr/pal/src/libunwind/include/libunwind-arm.h
@@ -280,6 +280,9 @@ unw_tdep_context_t;
   unsigned long *unw_base = unw_ctx->regs;                              \
   __asm__ __volatile__ (                                                \
     "mov r0, #0\n"                                                      \
-     "stmia %[base], {r0-r15}\n"                                         \
+     "stmia %[base]!, {r0-r15}\n"                                         \
+     "vstmia %[base], {d0-d15}\n" /* this also aligns return address to value stored by stmia */  \
-     "nop\n" /* align return address to value stored by stmia */         \
     : [r0] "=r" (r0) : [base] "r" (unw_base) : "memory");               \
@@ -295,6 +298,9 @@ unw_tdep_context_t;
     "nop\n"                                                             \
     ".code 32\n"                                                        \
     "mov r0, #0\n"                                                      \
     "stmia %[base]!, {r0-r14}\n"                                         \
     "adr r0, ret%=+1\n"                                                 \
-    "str r0, [%[base], #60]\n"                                          \
+    "stmia %[base]!, {r0}\n"                                         \
+    "vstmia %[base], {d0-d15}\n"                                        \
```

Reply:
Updated 5f2b78d93f201c5ca2ca9c24392311ae223b0cd7.

Reply:
The ! makes the stored value for [base] implementation-defined, but I don't think that matter (since the asm code clobbers it anyways). I guess I assumed from observation that R15 was defined to be PC+8 (eg as claimed here https://www.cc.gatech.edu/~hyesoon/spr10/lec_arm2.pdf)

Reply:
Oh, though you may also now need to update the register constraint to let the compiler know that you are now clobbering that register

Reply:
@janvorli, can we add a test case in Pri0 category for regular PRs that will:
* make use of arm32 fp registers and hit this path
* fail without applying your patch

I am not sure if it is as trivial as changing this test project\s property from 1 to 0 https://github.com/dotnet/runtime/blob/61b5c92c1509c9911eb3b3dd3fb684e48e717bc5/src/tests/JIT/IL_Conformance/Convert/TestConvertFromIntegral.csproj#L4? (as that was the project which was modified as part of the original fix https://github.com/dotnet/runtime/commit/1b5719c2e3dde393531eaeb5b5cde05dabeef5b8#diff-3df4332914893fb234891c86871f559e8d91d5b604df4073c56cdc36e9217b51)

Reply:
The thing is that I think we cannot enforce using non-volatile floating point registers (D8..D15) in the jitted code. Using  volatile ones doesn't cause issues, as we don't need to unwind it. Also, the register needs to be modified by the native code we unwind from to cause harm. Managed code uses the windows style unwinder, so it works unrelated to this change. 
We were hitting the issue in one of the tests with JIT stress enabled - the JIT stress resulted in using the D8 register. I think that might be just a coincidence though. I believe we run tests with jit stress enabled nightly or on some regular basis, so I'll check if that test still reproduces the problem so that we have some coverage.

Reply:
> Oh, though you may also now need to update the register constraint to let the compiler know that you are now clobbering that register

@vtjnash good point!

Reply:
@am11 I have verified that this update works with the TestConvertFromIntegral test by manually stepping through the unwinder.

Reply:
@janvorli, thank you for all the help. :)